Add remaining heatmaps

sjspielman opened the pull request:
Ever towards #516 

More heatmaps, and please enjoy the branch name, and if you don't enjoy the branch name, it means you still need to see _Zoolander_.


### Changes in this PR

1. Heatmap function has been implemented
2. Although I use the heatmap function for the first two heatmaps, I don't for the third (directly comparing singler-cellassign) since it is not a list of stacked heatmaps.
    - If we want, I can create a 2nd function to just make a heatmap from 1 matrix and call it from the current function. This would modularize the code further; let me know if we want to see this happen! I didn't do it (yet?) since this heatmap seemed sufficiently different that it might be fine to not use the function.
3. I realized that I have been developing this with unrealistic submitter labels - they have just been single letters! We need to allow for longer names, so I updated my test data to use long names to see what needed to change. Changes were:
    - I moved heatmap column labels to the bottom, but kept the column title at the top for spacing.
    - I wrote a function to more dynamically determine the right font size based on the label `nchar`
    - I allow for column labels to now be at an angle
4. Smaller change, I updated text to say "Jaccard ~similarity~ index" all around. Shorter and consistent with the wikipedia link.


For this round, I'm attaching the full report since there are really 3 heatmaps to look at given the changes I've made.
[celltypes_supplemental_report.html.zip](https://github.com/AlexsLemonade/scpca-nf/files/13228985/celltypes_supplemental_report.html.zip)

Again, we expect 1 more PR with textual updates (note I may bundle those with #514 text updates).

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
````diff
@@ -154,57 +253,18 @@ if (has_cellassign) {
-  heatmap_list, 
-  heatmap_legend_side = "bottom" 
+```{r, fig.height = heatmap_height, fig.width = 8.5, warning = FALSE}
+plot_heatmap(
+  jaccard_cluster_matrices, 
+  column_title = "Clusters", 
+  labels_font_size = find_label_size(jaccard_cluster_matrices)
 )
````
```suggestion
heatmap_list <- purrr::map(jaccard_cluster_matrices, \(jaccard_matrix, matrix_name) {
plot_heatmap(
  jaccard_matrix, 
  matrix_name,
  column_title = "Clusters", 
  labels_font_size = find_label_size(jaccard_cluster_matrices)
)} |> 
    # concatenate vertically into HeatmapList object
    purrr::reduce(ComplexHeatmap::`%v%`) 
  
  
  # Render the heatmap list
  ComplexHeatmap::draw(
    heatmap_list, 
    heatmap_legend_side = "bottom"

```
Then you can do this here. The only thing that I don't think this catches is dealing with the legend to keep. Can you go into the ComplexHeatmap object and remove legends like you can with ggplot?

jashapiro commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -94,6 +93,106 @@ make_jaccard_matrix <- function(celltype_df, colname1, colname2){
+# Set heatmap padding option
+ComplexHeatmap::ht_opt(TITLE_PADDING = grid::unit(0.2, "in"))
+
+#' Function to plot a ComplexHeatmap
```
Noting this is underselling! This function plots a list of matrices in a joined matrix!

sjspielman replied:
💪

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
````diff
@@ -226,7 +286,41 @@ knitr::asis_output(
+```
+
+```{r}
+# Calculate all jaccard matrices of interest for input to next heatmap
+jaccard_submitter_matrices <- list()
+
+n_celltypes <- 0 
+if (has_singler) {
+  jaccard_submitter_matrices[["SingleR"]] <- make_jaccard_matrix(
+    celltype_df, 
+    "submitter_celltype_annotation", 
+    "singler_celltype_annotation"
+  )
+  n_celltypes <- n_celltypes + length(unique(celltype_df$singler_celltype_annotation))
+}
+if (has_cellassign) {
+  jaccard_submitter_matrices[["CellAssign"]] <- make_jaccard_matrix(
+    celltype_df, 
+    "submitter_celltype_annotation", 
+    "cellassign_celltype_annotation"
+  )
+  n_celltypes <- n_celltypes + length(unique(celltype_df$cellassign_celltype_annotation))
+}
````
```suggestion
names <- c(dplyr::if_else(has_singler, "SingleR", NA),
           dplyr::if_else(has_cellassign, "CellAssign", NA)) |> 
  na.omit()

jaccard_submitter_matrices <- names |> 
  purrr::map( \(name) {
     lower_name <- stringr::str_to_lower(name)
    make_jaccard_matrix(celltype_df, "submitter_celltype_annotation", glue::glue("{lower_name}_celltype_annotation")} |> purrr::set_names(names)
```
I think you could simplify this so that you are not duplicating the call to calculating the jaccard matrix. 
Another thing to note here is that you will over write the `n_celltypes` variable if you have both SingleR and CellAssign. I think we want to handle that separately from calculating the jaccard matrix.

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
````diff
@@ -237,9 +331,58 @@ knitr::asis_output("
 
+```{r, eval = has_cellassign & has_singler, fig.height = heatmap_height, fig.width = 8.5}
+# Calculate jaccard matrix
+singler_cellassign_matrix <- make_jaccard_matrix(
````
Just note that if you take my suggestion about the function you will need to update this chunk to use that function.

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -94,6 +93,106 @@ make_jaccard_matrix <- function(celltype_df, colname1, colname2){
+        cluster_columns = FALSE,
+        column_title = column_title,
+        column_names_side = "bottom",
+        column_names_rot = column_names_angle,
```
Personally, I prefer a 90 degree label because I think it's easier to line up which cell type is in which column.

allyhawkins replied:
I think if you are worried about cell type names being too long, you could wrap them.

allyhawkins reviewed (commented):
My main comment here is that I think you should make the function just create one heatmap and then do any combining before drawing the heatmap outside of the function. The actual base of the heatmap is the same, the only difference is that we have stacked vs. not stacked. Because the stacked heatmap requires a list, it would be easy to create that list using `purrr::map` with a function that returns a single heatmap.

jashapiro commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -94,6 +93,106 @@ make_jaccard_matrix <- function(celltype_df, colname1, colname2){
+        ## Row parameters
+        cluster_rows = FALSE,
+        row_title = name, # each heatmap gets its own title
+        row_title_gp = grid::gpar(fontsize = 10), 
```
Can we set the `column_title_gp` to be the same? It is strange that they are different sizes. (They also seem to be padded differently, but maybe that is a function of the font size difference).

jashapiro reviewed (commented):
I think this looks good overall, but some comments below. 

* I think I would move the single heatmap function out to reuse as much as possible.
* I like "Jaccard similarity index" when we first describe it, but I am fine with "Jaccard index" for subsequent use. I don't think anyone will be confused.

sjspielman commented:
Next round:

- I wrote functions as requested, but in a slightly different way! There are now three functions:
  - 1 makes a standalone heatmap
  - 1 makes a list of heatmaps, in part by calling ☝️ 
  - 1 very simply _draws_ a heatmap
- I updated up some plot styling and cleaned up some other code spots as requested in review
- I'm now using @allyhawkins's suggested (and better!) `purrr::map()` approach for building up matrices

The one thing that I'm really not sure how to handle is dynamically setting the plot size itself. Stacked vs single heatmaps seem to have some difference in how height pans out, and in part because of this I can't tell whether the strategy I'm using is actually going to generalize well across reports. 
I'm thinking one way to handle the sizing could be to directly manipulate the heatmap dimensions itself and then set the plot size accordingly, but with some padding for vertical text? Blog post - https://jokergoo.github.io/2020/05/11/set-cell-width/height-in-the-heatmap/

Any thoughts on sizing are definitely appreciated! 

Report version as of now:
[celltypes_supplemental_report.html.zip](https://github.com/AlexsLemonade/scpca-nf/files/13243785/celltypes_supplemental_report.html.zip)

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -94,6 +93,147 @@ make_jaccard_matrix <- function(celltype_df, colname1, colname2){
+#'
+#' @param heatmap A heatmap or list of heatmaps to draw
+#' @param legend_side Side legend appears on, default bottom
+draw_heatmap <- function(heatmap, legend_side = "bottom") {
+  # Render the heatmap or list of heatmaps
+  ComplexHeatmap::draw(
+    heatmap, 
+    heatmap_legend_side = legend_side
+  )
+}
```
I'm not sure this really needs to be a function since you're just calling an already builtin function?

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -94,6 +93,147 @@ make_jaccard_matrix <- function(celltype_df, colname1, colname2){
+#' @param labels_font_size Font size to use for rows and column labels.
+#' @param keep_legend_name The name to use in the legend
+#' @param column_names_angle Angle for column names. Default is 0.
+#' @param col_fun Color function for the heatmap palette. Default is `heatmat_col_fun`.
```
```suggestion
#' @param col_fun Color function for the heatmap palette. Default is `heatmap_col_fun`.
```

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -94,6 +93,147 @@ make_jaccard_matrix <- function(celltype_df, colname1, colname2){
+      direction = "horizontal",
+      legend_width = unit(1.5, "in")
+    ),
+    # Retain only 1 legend in the final plot      
```
```suggestion
    # only keep legends that match `keep_legend_name`
```

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -94,6 +93,147 @@ make_jaccard_matrix <- function(celltype_df, colname1, colname2){
+#' @param col_fun Color function for the heatmap palette. Default is `heatmat_col_fun`.
+#'
+#' @return A list of ComplexHeatmap objects
+create_heatmap_list <- function(matrix_list,
+                              column_title, 
+                              labels_font_size,
+                              column_names_angle = 0,
+                              col_fun = heatmap_col_fun) {
```
```suggestion
create_heatmap_list <- function(matrix_list,
                                column_title, 
                                labels_font_size,
                                column_names_angle = 0,
                                col_fun = heatmap_col_fun) {
```
spacing

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -129,89 +277,37 @@ Cluster assignment was performed using the `r metadata(processed_sce)$cluster_al
+jaccard_cluster_matrices |>
+  create_heatmap_list(
+    column_title = "Clusters", 
+    labels_font_size = find_label_size(jaccard_cluster_matrices)
```
I wonder if you could use `purrr::flatten()` to get a list of all colnames in the list of data frames. Then you can use that as input to the `find_label_size` function and remove what's in the `if(is_list)` chunk and just start with a list of all column names.

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -226,20 +322,90 @@ knitr::asis_output(
+  "cellassign_celltype_annotation"
+)
+
+labels_font_size <- find_label_size(
+  c(
+    rownames(singler_cellassign_matrix), 
+    colnames(singler_cellassign_matrix)
+  ), 
+  is_list = FALSE
+)
```
```suggestion
labels_font_size <- find_label_size(
  c(
    rownames(singler_cellassign_matrix), 
    colnames(singler_cellassign_matrix)
  )
)
```
If you are able to figure out flattening the list as input, then you could remove the `is_list` here.

allyhawkins reviewed (commented):
Thanks for making these changes! Definitely easier to follow. I think I would still remove the `draw_heatmap` function since you are really just using an existing function there and not doing anything else special. 

Also in regards to heatmap size, I'm having trouble opening the zip file that you sent so can't comment on if what you're doing will turn out okay.

jashapiro reviewed (commented):
Getting there! I had a few additional comments, but largely agree with things Ally said, and also couldn't open the zip file.

sjspielman commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -226,20 +322,90 @@ knitr::asis_output(
+
+labels_font_size <- find_label_size(
+  c(
+    rownames(singler_cellassign_matrix), 
+    colnames(singler_cellassign_matrix)
```
Yeah, this clearly isn't right. I know I had a good reason for doing this at one point but it entirely escapes me now. Plus, this probably explains why I wasn't getting the "right dimensions" for this plot..!

Either way, I think we what _should_ be doing is considering:
1. The number of cell types along the y axis (row names)
2. The nchar of cell types along the x axis (column names)

jashapiro replied:
> Either way, I think we what _should_ be doing is considering:
> 
> 1. The number of cell types along the y axis (row names)
> 2. The nchar of cell types along the x axis (column names)

That sounds right to me. Maybe a little function to do the work, like the font size function, that takes a list of rownames and colnames and goes from there.

sjspielman commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -226,20 +322,90 @@ knitr::asis_output(
+
+# Set plot dimensions based on number of SingleR cell types
+n_celltypes <- length(levels(celltype_df$singler_celltype_annotation))
+heatmap_height <- n_celltypes/4 + 2 
```
No idea!!! For some reason 6 _really_ squished this plot in ways that it did not squish others, and 4 didn't. As noted in https://github.com/AlexsLemonade/scpca-nf/pull/536#issuecomment-1791424482 I'm not a fan of how I've set up sizing. Something is still missing. Let's circle back after I get the other reviews in.

sjspielman commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -129,89 +277,37 @@ Cluster assignment was performed using the `r metadata(processed_sce)$cluster_al
+  length()
+  
+# set height based on number of celltypes, and add on 2x padding
+# there are no long vertical layers here to accommodate
```
Ah, typo - "there are no long vertical **labels** here to accommodate" (cluster numbers along the axis, not long cell type names)

sjspielman commented:
Sorry for bad report! I think I just added a `.zip` extension instead of actually zipping it 😂 😭 
I'll post one when I re-request review next!

sjspielman commented:
I've gotten all reviews in, so next we just need to figure out the Right™️ way to determine heatmap height. Getting there!!

Here's a _confirmed-working_ `zip` file!
[celltypes_supplemental_report.html.zip](https://github.com/AlexsLemonade/scpca-nf/files/13253956/celltypes_supplemental_report.html.zip)

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -129,89 +256,38 @@ Cluster assignment was performed using the `r metadata(processed_sce)$cluster_al
+  
+# set height based on number of celltypes, and add on 2x padding
+# there are no long vertical labels here to accommodate, only cluster names (ints)
+heatmap_height <- length(unique_celltypes)/6 + 2*heatmap_padding
```
```suggestion
heatmap_height <- length(all_celltypes)/6 + 2*heatmap_padding
```
I don't know if this is the "right" way to do this, but I do think it's an okay way. If you are using cell types to determine the height then ideally, it should scale as the number of cell types change. You could try with adding in like 10 more submitter types and see if it still looks okay? But honestly it's probably fine and if we notice things being weird after running some samples we can always revisit this.

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -226,21 +302,85 @@ knitr::asis_output(
+  purrr::set_names(available_celltypes)
+
+# how many unique cell types?
+unique_celltypes <- jaccard_submitter_matrices |>
```
```suggestion
all_celltypes <- jaccard_submitter_matrices |>
```

allyhawkins commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -226,21 +302,85 @@ knitr::asis_output(
+# Set plot dimensions based on number of cell types plus:
+## 1x padding
+## 2" for long labels
+heatmap_height <- length(unique_celltypes)/6 + heatmap_padding + 2 
```
```suggestion
heatmap_height <- length(all_celltypes)/6 + heatmap_padding + 2 
```

allyhawkins reviewed (approved):
I think this looks good, and I'm okay with how you are determining the height. It's going to be hard to know for sure until we run a few samples and look at how the plots look once you change the number of cell types and clusters, etc.

sjspielman commented:
>  It's going to be hard to know for sure until we run a few samples and look at how the plots look once you change the number of cell types and clusters, etc.

I've been playing around with this for most of the morning so far, and I think you are almost certainly right that we'll need to circle back to this. Let me send over a function I've been working on just to make things more consistent, and I'll also open another issue to circle back and fix sizing once we have more (& realistic!) examples to work with.

sjspielman commented:
@allyhawkins I just pushed over the function I wrote which we can circle back to, for a quick peek from you?
Since my code diverged from what you reviewed, I swapped `unique_celltypes` -> `all_celltypes` on my own. 
One other change - use `stringr::str_length()` instead of `nchar()` just in case the vector that gets passed in is a factor and not a character vector.

All here: https://github.com/AlexsLemonade/scpca-nf/pull/536/commits/07bf56a4fe7dcfdda5fef4b2d9ed7b7b2b0a82b2

sjspielman commented:
Note - https://github.com/AlexsLemonade/scpca-nf/issues/541

jashapiro commented on templates/qc_report/celltypes_supplemental_report.rmd:
```diff
@@ -94,6 +93,157 @@ make_jaccard_matrix <- function(celltype_df, colname1, colname2){
+  keep_legend_name <- names(matrix_list)[1]
+
+  heatmap_list <- matrix_list |> 
+    purrr::imap(
+      create_single_heatmap, 
+      column_title, 
+      labels_font_size,
+      keep_legend_name,
+      column_names_angle = column_names_angle,
+      col_fun = col_fun
+    ) |>
```
Since the tidyverse people say they are now preferring lambda functions, I converted this, and put the `...` in for where I would do that. 
```suggestion
    purrr::imap(
      \(mat, name) create_single_heatmap(
        mat,
        name,
        column_title, 
        labels_font_size,
        keep_legend_name,
        col_fun,
        ...
      )
    ) |>
```

jashapiro reviewed (approved):
LGTM. I just had one real suggestion, which is to not to rename the ComplexHeatmaop `column_names_rot` variable in the wrapper functions. I even went so far as to suggest using `...` to pass it along, but that is up to you.

Also a couple more little future-proofing suggestions, but I don't think I should need to see this again.

allyhawkins reviewed (approved):
👍
